website/blog/2023-02-10.ferretdb-compare-arrays.md

```diff
@@ -0,0 +1,605 @@
+4. Compare next index, `1` and `'b'` using comparison order,
+   `1` is a number and less than string `'b'`.
+
+So `[3, 1, 'b']` is less than `[3, 'b', 'a']`.
+
+If an array has no more element to compare, the one with no more element is less.
+For the same reason empty array is less than arrays with elements.
+
+### Comparing Array and Other BSON Type
+When comparing an array with other BSON types, it doesn't *always* use comparison order of array BSON type.
+It often uses the greatest or the smallest value
+from the array to compare against other BSON type.
+
+In some cases, such as with the [`$max` and `$min` update operator](#max-min-update-operator), 
+the comparison order is used.
+In other situations, such as [sorting](#sorting), 
+the [`#gt` and `#lt` filter operators](#gt-lt-filter-operator),
+an element within the array is used to compare against other BSON type.
+
+[Sorting](#sorting)  used in `.sort({<<field>>: <<order>>})` uses following rules
+to compare array and other BSON types.
+
+* Ascending sort uses the smallest element of any BSON type from the array.
+* Descending sort uses the largest element of any BSON type from the array.
+* Empty array is less than null.
+
+[Filtering](#gt-lt-filter-operator) used in `$lt` and `$gt` operators use following rules
+to compare array and other BSON types.
+* `$lt` uses the smallest element of the same BSON type as non-array from the array.
+* `$gt` uses the largest element of the same BSON type as non-array from the array.
+* Empty array should not match other BSON type.
+
+[Updating](#max-min-update-operator) with `$max` and `$min` operators use following rules
+to compare array and other BSON types.
+* For array and other BSON type comparison, BSON comparison order is used.
+* Empty array is greater than null using BSON comparison order.
+
+#### Example of Using Ascending and Descending Sort
+How does array compare with other BSON type using one of the array element?
+In this example, `2` and `[3, 1]` are compared using ascending and descending sort.
+
+##### Array and Other BSON Type Ascending Sort
+Following steps use ascending sort to compare `2` and `[3, 1]`.
+
+1. It gets the smallest element of array `[3, 1]` which is `1`
+2. It compares `2` and `1`, `2` is greater than `1`.
+
+So `2` is greater than `[3, 1]` using ascending sort.
+
+##### Array and Other BSON Type Descending Sort
+Following steps use descending sort to compare `2` and `[3, 1]`.
+
+1. It gets the largest element of array `[3, 1]` which is `3`
+2. It compares `2` and `3`, `2` is less than `3`.
+
+So `2` is less than `[3, 1]` using descending sort.
+
+It might seem confusing at first, but ascending sort uses the smallest element
+from the array and descending sort uses the largest element from the array.
+
+For an empty array, it's a special case where empty array less than any BSON type.
+So `[]` is less than even `null` for [sorting](#comparison-in-sorting).
```

Comment:
```suggestion
##### Array and Other BSON Type Descending Sort

The following steps show the ascending sort comparison for `2` and `[3, 1]`:

1. It gets the smallest element of array `[3, 1]` which is `1`
2. It compares `2` and `1`, `2` is greater than `1`.

Therefore, when using ascending sort, `2` is greater than `[3, 1]`.

##### Array and Other BSON Type Descending Sort

The following steps show the descending sort comparison for  `2` and `[3, 1]`.

1. It gets the largest element of array `[3, 1]` which is `3`
2. It compares `2` and `3`, `2` is less than `3`.

Therefore, when using descending sort,  `2` is less than `[3, 1]`.

It might seem confusing at first, but note that ascending sort uses the smallest element
from the array, and descending sort uses the largest element from the array.

For an empty array, it's a special case; an empty array is less than any BSON type.
In that case, `[]` is less than even `null` for [sorting](#comparison-in-sorting).
```